server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -185,16 +235,234 @@ public InternalAggregation buildEmptyAggregation() {
+    private static class StandardOrderAggregator extends FiltersAggregator {
+        private final Query[] filters;
+        private Weight[] filterWeights;
+
+        private final int totalNumKeys;
+
+        public StandardOrderAggregator(
+            String name,
+            AggregatorFactories factories,
+            String[] keys,
+            Query[] filters,
+            boolean keyed,
+            String otherBucketKey,
+            SearchContext context,
+            Aggregator parent,
+            CardinalityUpperBound cardinality,
+            Map<String, Object> metadata
+        ) throws IOException {
+            super(name, factories, keys, keyed, otherBucketKey, context, parent, cardinality, metadata);
+            this.filters = filters;
+            if (otherBucketKey == null) {
+                this.totalNumKeys = keys.length;
+            } else {
+                this.totalNumKeys = keys.length + 1;
+            }
+        }
+
+        @Override
+        protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+            if (filterWeights == null) {
+                filterWeights = buildWeights(new MatchAllDocsQuery(), filters);
+            }
+            final Bits[] bits = new Bits[filters.length];
+            for (int i = 0; i < filters.length; ++i) {
+                bits[i] = Lucene.asSequentialAccessBits(ctx.reader().maxDoc(), filterWeights[i].scorerSupplier(ctx));
+            }
+            return new LeafBucketCollectorBase(sub, null) {
+                @Override
+                public void collect(int doc, long bucket) throws IOException {
+                    boolean matched = false;
+                    for (int i = 0; i < bits.length; i++) {
+                        if (bits[i].get(doc)) {
+                            collectBucket(sub, doc, bucketOrd(bucket, i));
+                            matched = true;
+                        }
+                    }
+                    if (otherBucketKey != null && false == matched) {
+                        collectBucket(sub, doc, bucketOrd(bucket, bits.length));
+                    }
+                }
+            };
+        }
+
+        final long bucketOrd(long owningBucketOrdinal, int filterOrd) {
+            return owningBucketOrdinal * totalNumKeys + filterOrd;
+        }
+
+        @Override
+        public boolean collectsInFilterOrder() {
+            return false;
+        }
+    }
+
+    protected Weight[] buildWeights(Query topLevelQuery, Query filters[]) throws IOException{
+        Weight[] weights = new Weight[filters.length];
+        for (int i = 0; i < filters.length; ++i) {
+            Query filter = filterMatchingBoth(topLevelQuery, filters[i]);
+            weights[i] = context.searcher().createWeight(context.searcher().rewrite(filter), ScoreMode.COMPLETE_NO_SCORES, 1);
+        }
+        return weights;
+    }
+
+    private Query filterMatchingBoth(Query lhs, Query rhs) {
```

Comment:
This method and everything in it is kind of shameful but it gives a 2x speed improvement.

Reply:
:+1: